### ISSUE TITLE
doAs method in Ranger plugin incompatible with Trino’s Hadoop library - Trino 467

### ISSUE THREAD

**Issue by nukima** [trinodb/trino#24498]
**Description:**

When configuring the Ranger plugin for Trino 467 with the `ranger.hadoop.config.resource` property, Trino fails to start due to a `NoSuchMethodError` in the `doAs` method of the `UserGroupInformation` class. The issue appears to arise because the specified `core-site.xml` contains Kerberos-related properties, such as:
```xml
<property>
    <name>hadoop.security.authentication</name>
    <value>kerberos</value>
</property>

``` 
This configuration sets `isSecureMode=True`, although Trino itself is not configured to use Kerberos authentication.

**Steps to Reproduce:**

- Set up Trino 467 and enable the Apache Ranger plugin.
- Add the `ranger.hadoop.config.resource` property to Trino's configuration, pointing to a core-site.xml file that includes Kerberos settings (e.g., `hadoop.security.authentication=kerberos`).
- Start the Trino server.

**Observed Behavior:**

Trino fails to start, and the logs contain the following error:
```log
Caused by: java.lang.NoSuchMethodError: 'java.lang.Object org.apache.hadoop.security.UserGroupInformation.doAs(java.security.PrivilegedAction)'
	at org.apache.ranger.admin.client.RangerAdminRESTClient.getRangerRolesDownloadResponse(RangerAdminRESTClient.java:1315)
	at org.apache.ranger.admin.client.RangerAdminRESTClient.getRolesIfUpdatedWithCred(RangerAdminRESTClient.java:1183)
	at org.apache.ranger.admin.client.RangerAdminRESTClient.getRolesIfUpdated(RangerAdminRESTClient.java:148)
	at org.apache.ranger.plugin.util.RangerRolesProvider.loadUserGroupRolesFromAdmin(RangerRolesProvider.java:172)
	at org.apache.ranger.plugin.util.RangerRolesProvider.loadUserGroupRoles(RangerRolesProvider.java:112)
	at org.apache.ranger.plugin.util.PolicyRefresher.loadRoles(PolicyRefresher.java:563)
	at org.apache.ranger.plugin.util.PolicyRefresher.startRefresher(PolicyRefresher.java:138)
	at org.apache.ranger.plugin.service.RangerBasePlugin.init(RangerBasePlugin.java:254)
	at io.trino.plugin.ranger.RangerSystemAccessControl.<init>(RangerSystemAccessControl.java:159)
	at io.trino.plugin.ranger.RangerSystemAccessControl$$FastClassByGuice$$1b4469.GUICE$TRAMPOLINE(<generated>)
	at io.trino.plugin.ranger.RangerSystemAccessControl$$FastClassByGuice$$1b4469.apply(<generated>)
	at com.google.inject.internal.DefaultConstructionProxyFactory$FastClassProxy.newInstance(DefaultConstructionProxyFactory.java:82)
	at com.google.inject.internal.ConstructorInjector.provision(ConstructorInjector.java:114)
	at com.google.inject.internal.ConstructorInjector.access$000(ConstructorInjector.java:33)
	at com.google.inject.internal.ConstructorInjector$1.call(ConstructorInjector.java:98)
	at com.google.inject.internal.ProvisionListenerStackCallback$Provision.provision(ProvisionListenerStackCallback.java:109)
	at io.airlift.bootstrap.LifeCycleModule.provision(LifeCycleModule.java:53)
	at com.google.inject.internal.ProvisionListenerStackCallback$Provision.provision(ProvisionListenerStackCallback.java:117)
	at com.google.inject.internal.ProvisionListenerStackCallback.provision(ProvisionListenerStackCallback.java:66)
	at com.google.inject.internal.ConstructorInjector.construct(ConstructorInjector.java:93)
	at com.google.inject.internal.ConstructorBindingImpl$Factory.get(ConstructorBindingImpl.java:300)
	at com.google.inject.internal.ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:40)
	at com.google.inject.internal.SingletonScope$1.get(SingletonScope.java:169)
	at com.google.inject.internal.InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:45)
	at com.google.inject.internal.InternalInjectorCreator.loadEagerSingletons(InternalInjectorCreator.java:213)
	at com.google.inject.internal.InternalInjectorCreator.injectDynamically(InternalInjectorCreator.java:186)
	... 19 more
``` 
[Full log here: trino-server-fail.log](https://github.com/user-attachments/files/18159462/trino-server-fail.log)

**Analysis**:

- Although Trino's authentication is not configured with Kerberos, specifying a core-site.xml configuration containing `hadoop.security.authentication=kerberos` causes the Hadoop library to enter secure mode (`isSecureMode=True`).
- The Ranger plugin (`RangerAdminRESTClient.getRangerRolesDownloadResponse` - `org.apache.ranger:ranger-plugins-common:2.5.0`) uses the `doAs` method of the `UserGroupInformation` class with an argument of type `PrivilegedAction<ClientResponse>`.
- The `UserGroupInformation` class in the Hadoop library (`io.trino.hadoop:-hadoop-apache:3.3.5-3`) only defines a `doAs` method that accepts a `PrivilegedExceptionAction<T>`.
- This mismatch leads to the `NoSuchMethodError`.

**Environment Details:**

Trino version: 467
Ranger plugin version: org.apache.ranger:ranger-plugins-common:2.5.0
Hadoop library: io.trino.hadoop:-hadoop-apache:3.3.5-3

**Workaround:**

N/A. The issue requires code changes in the Ranger plugin.

**Impact:**

This issue blocks the integration of Trino 467 with Apache Ranger when the `ranger.hadoop.config.resource` property is configured to point to a Kerberos-enabled core-site.xml.

The issue was identified by my colleague (anh Tung) but they prefer not to be mentioned in this ticket.

**Comment by mosabua** [comment]
From what I know the current Ranger plugin does support Kerberos authentication - so this is not really a bug or issue.. more a feature request.

Can you confirm that it works if you remove the Kerberos stuff in the xml file?

Maybe you can send a PR .. or see if you can work with @mneethiraj and others to implement a fix or work towards Kerberos support.

**Comment by nukima** [comment]
Thanks for the feedback. I can confirm that removing the `ranger.hadoop.config.resource` configuration resolves the startup issue. However, the problem appears to be more nuanced than a simple feature request.

The root cause is a method signature mismatch in the Ranger plugin when interacting with the Hadoop UserGroupInformation class. Specifically:
- The Ranger plugin (ranger-plugins-common:2.5.0) attempts to use `doAs(PrivilegedAction)` 
- The Hadoop library in Trino (hadoop-apache:3.3.5-3) only defines `doAs(PrivilegedExceptionAction)`
This leads to a `NoSuchMethodError` when Trino starts.

I'd be happy to work with the maintainers to investigate a potential fix or compatibility adjustment.

**Comment by nukima** [comment]
@mneethiraj Could you share your thoughts on this issue? What steps could we take to address it? Happy to contribute if needed.

**Comment by mneethiraj** [comment]
@nukima  - thank you for reporting the issue and details. This is tracked in Apache Ranger via [RANGER-4891](https://issues.apache.org/jira/browse/RANGER-4891); the fix is already merged in master and ranger-2.6 branches and will be included in the next maintenance release.

**Comment by nukima** [comment]
Thank you @mneethiraj for the update! I'll keep an eye out for the next maintenance release. Appreciate the quick response!
